README.md

```diff
@@ -140,6 +142,106 @@ title,content
 My Title,Hello World!
 ```
 
+### Creating a custom Task
+
+TODO: Add generation instructions
+
+If you have a special use case requiring iteration over an unsupported
+collection type, such as external resources fetched from some API, you can
+implement the `enumerator_builder` method instead.
+
+This method should return an object responding to `enumerator(context:)` with
+an Enumerator, yielding pairs of `[item, item_cursor]` to the `process` method
+your Task defines. In order for your enumerator to support resuming iteration
+part way through, you may use the `context.cursor`.
+
+You may optionally provide an implementation for `count`, if appropriate.
+
+```ruby
+# app/tasks/maintenance/shopping_list_task.rb
+module Maintenance
+  class ShoppingListTask < MaintenanceTasks::Task
+    def enumerator_builder
+      IngredientsEnumerator.new
+    end
+
+    def process(ingredient)
+      ShoppingList.add(ingredient)
+    end
+
+    class IngredientEnumerator
+      def enumerator(context:)
+        Enumerator.new do |yielder|
+          cursor = context.cursor
+
+          if cursor.nil?
+            recipe = FancyRecipeAPI.random_recipe
+            ingredient_id = nil
+          else
+            recipe_id, ingredient_id = cursor.split(':', 2)
+            recipe = FancyRecipeAPI.recipe(recipe_id)
+          end
+
+          loop do
+            page = if ingredient_id.nil?
+              FancyRecipeAPI
+                .paginated_ingredients(recipe.id, max: 5)
+            else
+              FancyRecipeAPI
+                .paginated_ingredients(recipe.id, max: 5, after: ingredient_id)
+            end
+
+            page.entries.each do |ingredient|
+              ingredient_id = ingredient.id
+              cursor = "#{recipe.id}:#{ingredient.id}"
+
+              yielder.yield([ingredient, cursor])
+            end
+
+            break unless page.has_next?
+          end
+        end
+      end
+    end
+  end
+end
+```
+
+In some cases, you may have no use for a cursor (e.g. taking items off a queue
+until it is empty), in which case your Enumerator may yield `nil`
+cursors (i.e. pairs of `[item, nil]`).
```

Comment:
Wouldn't that work by simply yielding the item by itself?
```ruby
foo = Enumerator.new { |y| y.yield :item }
# => #<Enumerator: #<Enumerator::Generator:0x00007faf8dac0b90>:each>
foo.each do |item, cursor| p item, cursor end
# :item
# nil
# => [:item, nil]
```

Reply:
Yeah, we identified both
- wanting to have consumers be explicit about not having a cursor, as well as
- the existence of a "footgun", in the case where the `element` they want to yield is an array
  (say they're implementing a cursorless custom batched task, or forget to include a cursor for a custom task yielding pairs)

```ruby
def example(&block)
  Enumerator.new(&block).each do |element, cursor|
    puts "element: #{element.inspect.ljust(16)} cursor: #{cursor.inspect}"
  end
end

example { |y| y.yield :item, :cursor }              # element: :item            cursor: :cursor
example { |y| y.yield :item, nil }                  # element: :item            cursor: nil
example { |y| y.yield :item }                       # element: :item            cursor: nil

example { |y| y.yield [:item, :cursor] }            # element: :item            cursor: :cursor
example { |y| y.yield [:item, nil] }                # element: :item            cursor: nil
example { |y| y.yield [:item] }                     # element: :item            cursor: nil

example { |y| y.yield [[:pair1, :pair2], :cursor] } # element: [:pair1, :pair2] cursor: :cursor
example { |y| y.yield [[:pair1, :pair2], nil] }     # element: [:pair1, :pair2] cursor: nil
example { |y| y.yield [[:pair1, :pair2]] }          # element: [:pair1, :pair2] cursor: nil

example { |y| y.yield [:pair1, :pair2], :cursor }   # element: [:pair1, :pair2] cursor: :cursor
example { |y| y.yield [:pair1, :pair2], nil }       # element: [:pair1, :pair2] cursor: nil
example { |y| y.yield [:pair1, :pair2] }            # element: :pair1           cursor: :pair2
                                                    # cursor forgotten or omitted?      ^^^^^^
```

Without an explicit cursor, we're unable to distinguish between yielding an array, or yielding and item and cursor. This occurs for any case where they array has more than one element:

```ruby
example { |y| y.yield [:pair1, :pair2, :pair3] }    # element: :pair1           cursor: :pair2
                                                    # cursor forgotten or omitted?      ^^^^^^
```

Reply:
That said, I'm not sure how to implement that restriction, given the way block variables are auto-unpacked, so we might have to leave it. 🤷

Reply:
Let's leave it, we'll make things clear in the documentation!